### PR TITLE
feature/PATRO23-370-refresh-token

### PR DESCRIPTION
**Pull request by Kepeczsz**
https://tracker.intive.com/jira/browse/PATRO23-370

### CODE REVIEW

File: src/modules/user/application/RefreshingUserToken/RefreshUserToken.cs [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+/// Record which holds user refresh token.
+/// </summary>
+/// <param name="RefreshToken">refresh token.</param>
+public record RefreshUserToken(string RefreshToken) : IQuery<AccesToken>;
+
+/// <summary>
+/// Record which holds newly generated access token.
+/// </summary>
+/// <param name="AccessToken"> newly generated access token.</param>
+public record AccesToken(string AccessToken);
```
Comment (IzabelaPalubicka) [inline src/modules/user/application/RefreshingUserToken/RefreshUserToken.cs:17]: typo  `AccessToken`

**Review by IzabelaPalubicka — APPROVED** [review]

**Review by AlanP22 — APPROVED** [review]

File: src/modules/user/application/RefreshingUserToken/RefreshUserToken.cs [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+		var accessToken = jwtHandler.ReadJwtToken(oldAccessToken);
+		var refreshToken = jwtHandler.ReadJwtToken(command.RefreshToken);
+
+		if (refreshToken.ValidTo < DateTime.UtcNow)
+		{
+			throw new AppException("refresh token is expired.");
+		}
+
+		if (accessToken.ValidTo.AddMinutes(-1) <= DateTime.UtcNow)
```
Comment (mateuszgiedzinski) [inline src/modules/user/application/RefreshingUserToken/RefreshUserToken.cs:48]: Don't use DateTime.UtcNow directly. Use time service instead. Also you can move that logic to extension and change this to accessToken.IsExpired() extension method.

File: src/modules/user/application/RefreshingUserToken/RefreshUserToken.cs [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+	{
+		string oldAccessToken = await this.keycloakService.ExtractAccessTokenFromClientToken(cancellationToken);
+		string response = string.Empty;
+
+		var jwtHandler = new JwtSecurityTokenHandler();
+
+		var accessToken = jwtHandler.ReadJwtToken(oldAccessToken);
+		var refreshToken = jwtHandler.ReadJwtToken(command.RefreshToken);
+
+		if (refreshToken.ValidTo < DateTime.UtcNow)
```
Comment (mateuszgiedzinski) [inline src/modules/user/application/RefreshingUserToken/RefreshUserToken.cs:43]: You can move that logic to extension and change this to refreshToken.IsExpired() using  extension method.

File: src/modules/user/application/RefreshingUserToken/RefreshUserToken.cs (hunk 2 above)
Comment (mateuszgiedzinski) [reply src/modules/user/application/RefreshingUserToken/RefreshUserToken.cs:48]: IsExpired can have some TimeSpan parameter by default set to 0 minutes but in this case can be -1.minutes. 
you can use that parameter as margin to check when token is treated as expired.

**Review by mateuszgiedzinski — CHANGES_REQUESTED** [review]

File: src/modules/user/api/Controllers/UserController.cs [hunk 4]
```diff
@@ -110,6 +111,21 @@ public async Task<IActionResult> SignUp([FromBody] CreateUser command)
...
+	/// </summary>
+	/// <param name="query">The refresh token command, which includes refresh token.</param>
+	/// <returns>New access token.</returns>
+	[ProducesResponseType(StatusCodes.Status200OK)]
+	[ProducesResponseType(StatusCodes.Status400BadRequest)]
+	[AllowAnonymous]
+	[HttpPost("refresh-token")]
+	public async Task<IActionResult> RefreshToken([FromBody] RefreshUserToken query)
+	{
+		var reponse = await this.queryBus.Query<RefreshUserToken, AccessUserToken>(query);
```
Comment (mateuszgiedzinski) [inline src/modules/user/api/Controllers/UserController.cs:125]: I don't see any validation for input parrameters.

**Review by mateuszgiedzinski — CHANGES_REQUESTED** [review]

**Review by mateuszgiedzinski — APPROVED** [review]

**Review by mateuszmuszynski — APPROVED** [review]

**Pull request merged**